File: easy-permission/src/main/java/com/github/boybeak/easypermission/EasyPermission.java
package com.github.boybeak.easypermission;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public class EasyPermission {

    private static final String KEY_ = "com.github.boybeak.easypermission.";
    public static final String KEY_ID = KEY_ + "ID", KEY_PERMISSION_LIST = KEY_ + "PERMISSION_LIST",
            KEY_REQUEST_CODE = KEY_ + "REQUEST_CODE";
    public static final int REQUEST_CODE_DEFAULT = 127;

    private static final String TAG_PERMISSION_FRAGMENT = "com.github.boybeak.easypermission.TAG_PERMISSION_FRAGMENT";

    private static Map<String, Callback> sKeyCallbackMap = new HashMap<>();

    private static void actionGranted(String id, int requestCode, List<String> permissions) {
        Callback callback = sKeyCallbackMap.get(id);
        if (callback != null) {
            callback.onGranted(permissions, requestCode);
        }
    }

    private static void actionDenied(String id, int requestCode, String permission, boolean neverAsk) {
        Callback callback = sKeyCallbackMap.get(id);
        if (callback != null) {
            callback.onDenied(permission, requestCode, neverAsk);
        }
    }

    static void actionPermissionResult(String id, Activity activity, int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        boolean result = true;
        boolean neverAsk = false;
        String permission = null;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            for (int i = 0; i < grantResults.length; i++) {
                int r = grantResults[i];
                result = r == PackageManager.PERMISSION_GRANTED;
                if (!result) {
                    permission = permissions[i];
                    neverAsk = !ActivityCompat.shouldShowRequestPermissionRationale(activity, permission);
                    break;
                }
            }
        }
        if (result) {
            actionGranted(id, requestCode, Arrays.asList(permissions));
        } else {
            actionDenied(id, requestCode, permission, neverAsk);
        }
        releaseCallback(id);
    }

    public static boolean isPermissionGranted(Context context, String ... permissions) {
        return isPermissionGranted(context, Arrays.asList(permissions));
    }

    public static boolean isPermissionGranted(Context context, List<String> permissions) {
        for (String p : permissions) {
            int result = ContextCompat.checkSelfPermission(context, p);
            if (result != PackageManager.PERMISSION_GRANTED) {
                return false;
            }
        }
        return true;
    }

    private static void releaseCallback(String id) {
        sKeyCallbackMap.remove(id);
    }

    public static EasyPermission ask(String ... permissions) {
        return ask(Arrays.asList(permissions));
    }

    public static EasyPermission ask(List<String> permissions) {
        if (permissions.isEmpty()) {
            throw new IllegalArgumentException("permission list should not be empty");
        }
        EasyPermission ph = new EasyPermission();
        ph.permissions.addAll(permissions);
        return ph;
    }

    private ArrayList<String> permissions = new ArrayList<>();

    private EasyPermission() {

    }

    public void go(Context context, Callback callback) {
        go(context, REQUEST_CODE_DEFAULT, callback);
    }

    public void go(Context context, int requestCode, Callback callback) {

        if (EasyPermission.isPermissionGranted(context, permissions)) {
            callback.onGranted(permissions, requestCode);
            return;
        }

        String id = UUID.randomUUID().toString();

        if (context instanceof FragmentActivity) {
            FragmentActivity fa = (FragmentActivity)context;
            FragmentManager fm = fa.getSupportFragmentManager();
            PermissionAgentFragment agentFragment = (PermissionAgentFragment)fm.findFragmentByTag(TAG_PERMISSION_FRAGMENT);
            if (agentFragment == null) {
                agentFragment = new PermissionAgentFragment();
                agentFragment.setPermissionsArgs(id, requestCode, permissions);
                fm.beginTransaction().add(agentFragment, TAG_PERMISSION_FRAGMENT)
                        .commitNowAllowingStateLoss();
            } else {
                String[] permissionArray = new String[permissions.size()];
                permissions.toArray(permissions.toArray());
                agentFragment.requestPermissions(permissionArray, requestCode);
            }
        } else {
            Intent it = new Intent(context, PermissionAgentActivity.class)
                    .putExtra(KEY_ID, id)
                    .putStringArrayListExtra(KEY_PERMISSION_LIST, permissions)
                    .putExtra(KEY_REQUEST_CODE, requestCode);
            context.startActivity(it);
        }

        sKeyCallbackMap.put(id, callback);

    }

}

File: easy-permission/src/main/java/com/github/boybeak/easypermission/ext/_EasyPermission.kt
package com.github.boybeak.easypermission.ext

import android.content.Context
import androidx.core.content.PermissionChecker
import androidx.fragment.app.Fragment
import com.github.boybeak.easypermission.Callback
import com.github.boybeak.easypermission.EasyPermission

fun Context.withPermission(
    requestCode: Int = EasyPermission.REQUEST_CODE_DEFAULT,
    permission: String,
    onGranted: (permissions: List<String>) -> Unit
) {
    withPermission(requestCode, permission, onGranted, null)
}

fun Context.withPermission(
    requestCode: Int = EasyPermission.REQUEST_CODE_DEFAULT,
    permission: String,
    onGranted: (permissions: List<String>) -> Unit,
    onDenied: ((permission: String, neverAsk: Boolean) -> Unit)?
) {
    withPermissions(requestCode, permission, onGranted = onGranted, onDenied = onDenied)
}

fun Context.withPermissions(
    requestCode: Int = EasyPermission.REQUEST_CODE_DEFAULT,
    vararg permissions: String,
    onGranted: (permissions: List<String>) -> Unit
) {
    withPermissions(requestCode, *permissions, onGranted = onGranted, onDenied = null)
}

fun Context.withPermissions(
    requestCode: Int = EasyPermission.REQUEST_CODE_DEFAULT,
    vararg permissions: String,
    onGranted: (permissions: List<String>) -> Unit,
    onDenied: ((permission: String, neverAsk: Boolean) -> Unit)?
) {
    EasyPermission.ask(*permissions).go(this, requestCode, object : Callback {
        override fun onGranted(permissions: MutableList<String>, requestCode: Int) {
            onGranted.invoke(permissions)
        }

        override fun onDenied(
            permission: String,
            requestCode: Int,
            neverAsk: Boolean
        ) {
            onDenied?.invoke(permission, neverAsk)
        }
    })
}

fun Context.withPermissionList(
    requestCode: Int = EasyPermission.REQUEST_CODE_DEFAULT,
    permissions: List<String>,
    onGranted: (permissions: List<String>) -> Unit
) {
    withPermissionList(requestCode, permissions, onGranted, null)
}

fun Context.withPermissionList(
    requestCode: Int = EasyPermission.REQUEST_CODE_DEFAULT,
    permissions: List<String>,
    onGranted: (permissions: List<String>) -> Unit,
    onDenied: ((permission: String, neverAsk: Boolean) -> Unit)?
) {

    EasyPermission.ask(permissions).go(this, requestCode, object : Callback {
        override fun onGranted(permissions: MutableList<String>, requestCode: Int) {
            onGranted.invoke(permissions)
        }

        override fun onDenied(
            permission: String,
            requestCode: Int,
            neverAsk: Boolean
        ) {
            onDenied?.invoke(permission, neverAsk)
        }
    })
}

fun Fragment.withPermissionList(
    requestCode: Int = EasyPermission.REQUEST_CODE_DEFAULT,
    permissions: List<String>,
    onGranted: (permissions: List<String>) -> Unit,
    onDenied: ((permission: String, neverAsk: Boolean) -> Unit)?
) {
    requireContext().withPermissionList(requestCode, permissions, onGranted, onDenied)
}

fun Fragment.withPermissionList(
    requestCode: Int = EasyPermission.REQUEST_CODE_DEFAULT,
    permissions: List<String>,
    onGranted: (permissions: List<String>) -> Unit
) {
    requireContext().withPermissionList(requestCode, permissions, onGranted)
}

fun Fragment.withPermission(
    requestCode: Int = EasyPermission.REQUEST_CODE_DEFAULT,
    permission: String,
    onGranted: (permissions: List<String>) -> Unit
) {
    requireContext().withPermission(requestCode, permission, onGranted)
}

fun Fragment.withPermission(
    requestCode: Int = EasyPermission.REQUEST_CODE_DEFAULT,
    permission: String,
    onGranted: (permissions: List<String>) -> Unit,
    onDenied: ((permission: String, neverAsk: Boolean) -> Unit)?
) {
    requireContext().withPermission(requestCode, permission, onGranted, onDenied)
}

fun Fragment.withPermissions(
    requestCode: Int = EasyPermission.REQUEST_CODE_DEFAULT,
    vararg permissions: String,
    onGranted: (permissions: List<String>) -> Unit
) {
    requireContext().withPermissions(requestCode, *permissions, onGranted = onGranted)
}

fun Fragment.withPermissions(
    requestCode: Int = EasyPermission.REQUEST_CODE_DEFAULT,
    vararg permissions: String,
    onGranted: (permissions: List<String>) -> Unit,
    onDenied: ((permission: String, neverAsk: Boolean) -> Unit)?
) {
    requireContext().withPermissions(requestCode, *permissions, onGranted = onGranted, onDenied = onDenied)
}

fun Context.isPermissionsGranted(vararg permissions: String): Boolean {
    return EasyPermission.isPermissionGranted(this, *permissions)
}

fun Context.isPermissionsGranted(permissions: List<String>): Boolean {
    return EasyPermission.isPermissionGranted(this, permissions)
}